core/src/test/java/org/elasticsearch/snapshots/SnapshotSharedTest.java

```diff
@@ -0,0 +1,239 @@
+import static org.hamcrest.Matchers.greaterThan;
+import static org.junit.Assert.assertThat;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import org.elasticsearch.Version;
+import org.elasticsearch.action.ListenableActionFuture;
+import org.elasticsearch.action.admin.cluster.snapshots.create.CreateSnapshotResponse;
+import org.elasticsearch.action.admin.cluster.snapshots.restore.RestoreSnapshotResponse;
+import org.elasticsearch.action.admin.cluster.snapshots.status.SnapshotIndexShardStatus;
+import org.elasticsearch.action.admin.cluster.snapshots.status.SnapshotStatus;
+import org.elasticsearch.action.admin.indices.flush.FlushResponse;
+import org.elasticsearch.action.index.IndexRequestBuilder;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.metadata.IndexMetaData;
+import org.elasticsearch.common.logging.ESLogger;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.ByteSizeUnit;
+import org.elasticsearch.test.ESIntegTestCase;
+import org.hamcrest.Matcher;
+
+/**
+ * Test cases shared between SharedClusterSnapshotRestoreIT and SnapshotBackwardsCompatibilityIT.
+ */
+public class SnapshotSharedTest {
+    /**
+     * Callback used after snapshots have been created.
+     */
+    public static interface AfterSnapshotAction {
+        public static AfterSnapshotAction NOOP = new AfterSnapshotAction() {
+            @Override
+            public void run(String[] indices, int numDocs) {
+            }
+        };
+        /**
+         * Called after the snapshots have been created.
+         * @param indices the indices that the test is using
+         * @param numDocs the number of docs in open indices on the cluster
+         */
+        public void run(String[] indices, int numDocs);
+    }
+
+    public static void testBasicWorkflow(ESLogger logger, ESIntegTestCase testCase, AfterSnapshotAction afterSnapshotAction, Matcher<Version> snapshotVersion) {
+        logger.info("-->  creating repository");
+        assertAcked(client().admin().cluster().preparePutRepository("test-repo")
+                .setType("fs").setSettings(Settings.settingsBuilder()
+                        .put("location", randomRepoPath())
+                        .put("compress", randomBoolean())
+                        .put("chunk_size", randomIntBetween(100, 1000), ByteSizeUnit.BYTES)));
+
+        testCase.createIndex("test-idx-1", "test-idx-2", "test-idx-3");
+        testCase.ensureGreen();
+
+        logger.info("--> indexing some data");
+        for (int i = 0; i < 100; i++) {
+            testCase.index("test-idx-1", "doc", Integer.toString(i), "foo", "bar" + i);
+            testCase.index("test-idx-2", "doc", Integer.toString(i), "foo", "baz" + i);
+            testCase.index("test-idx-3", "doc", Integer.toString(i), "foo", "baz" + i);
+        }
+        testCase.refresh();
+        assertHitCount(client().prepareCount("test-idx-1").get(), 100L);
```

Comment:
`#prepareCount()`is deprecated. Do you think it makes sense to correct it already in this PR?